File: csharp/Bingo.Test/BingoTest.cs
namespace Bingo.Test;

/**
 * This class is identical to {@link BingoTest} but all BDD-style methods are
 * inlined (and thus this test does depend at multiple points on the
 * implementation details of {@link BingoBoard})
 */
public class BingoTest
{

    BingoBoard board;

    [Fact]
    void AnNewlyCreatedBoardIsNotInitialized()
    {
        board = new BingoBoard(1, 1);
        board.IsInitialized().Should().BeFalse();
    }

    [Fact]
    void WhenAllFieldsAreSetTheBoarIsInitialized()
    {
        var anyValue = "42";
        board = new BingoBoard(1, 1);
        board.DefineCell(0, 0, anyValue);
        board.IsInitialized().Should().BeTrue();
    }

    [Fact]
    void WhenAllFieldsOnRectangularBoardAreSetItIsInitialized()
    {
        var one = "one, two, three";
        var two = "Bingo cells can contain any text";
        board = new BingoBoard(1, 2);
        board.DefineCell(0, 0, one);
        board.DefineCell(0, 1, two);
        board.IsInitialized().Should().BeTrue();
    }

    [Fact]
    void ADefinedCellCantBeRedefinedEvenIfItsTheSameValue()
    {
        var anyValue = "42";
        board = new BingoBoard(1, 1);
        board.DefineCell(0, 0, anyValue);
        Action secondCall = () => board.DefineCell(0, 0, anyValue);
        secondCall.Should()
            .Throw<InvalidOperationException>()
            .Where(e => e.Message.Contains("already defined"));
    }

    [Fact]
    void DuplicateCellsAreNotAllowed()
    {
        var anyValue = "42";
        board = new BingoBoard(2, 2);
        board.DefineCell(0, 1, anyValue);
        Action secondCall = () => board.DefineCell(1, 0, anyValue);
        secondCall.Should()
            .Throw<InvalidOperationException>()
            .WithMessage(anyValue + " already present at 0,1");
    }

    [Fact]
    void ANonInitializedBoardCannotBeMarked()
    {
        board = new BingoBoard(1, 1);
        Action markCell = () => board.MarkCell(0, 0);
        markCell.Should()
            .Throw<InvalidOperationException>()
            .Where(e => e.Message.EndsWith("not initialized"));
    }

    [Fact]
    void WhenAllCellGetsMarkedItIsMarked()
    {
        var anyValue = "42";
        board = new BingoBoard(1, 1);
        board.DefineCell(0, 0, anyValue);
        board.MarkCell(0, 0);
        board.IsMarked(0, 0).Should().BeTrue();
    }

}

File: php/tests/BingoTest.php
<?php

namespace KataTest;

use Kata\BingoBoard;
use PHPUnit\Framework\TestCase;

class BingoTest extends TestCase
{
    private BingoBoard $board;

    public function testAnNewlyCreatedBoardIsNotInitialized(): void
    {
        $this->board = new BingoBoard(1, 1);
        $this->assertFalse($this->board->isInitialized());
    }

    public function testWhenAllFieldsAreSetTheBoardIsInitialized(): void
    {
        $anyValue = "42";
        $this->board = new BingoBoard(1, 1);
        $this->board->defineCell(0, 0, $anyValue);
        $this->assertTrue($this->board->isInitialized());
    }

    public function testWhenAllFieldsOnRectangularBoardAreSetItIsInitialized(): void
    {
        $one = "one, two, three";
        $two = "Bingo cells can contain any text";
        $this->board = new BingoBoard(1, 2);
        $this->board->defineCell(0, 0, $one);
        $this->board->defineCell(0, 1, $two);
        $this->assertTrue($this->board->isInitialized());
    }

    public function testADefinedCellCantBeRedefinedEvenIfItsTheSameValue(): void
    {
        $anyValue = "42";
        $this->board = new BingoBoard(1, 1);
        $this->board->defineCell(0, 0, $anyValue);

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessageMatches('/already defined/');
        $this->board->defineCell(0, 0, $anyValue);
    }

    public function testDuplicateCellsAreNotAllowed(): void
    {
        $anyValue = "42";
        $this->board = new BingoBoard(2, 2);
        $this->board->defineCell(0, 1, $anyValue);

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessageMatches('/' . preg_quote($anyValue . " already present at 0,1") . '/');
        $this->board->defineCell(1, 0, $anyValue);
    }

    public function testANonInitializedBoardCannotBeMarked(): void
    {
        $this->board = new BingoBoard(1, 1);

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessageMatches('/not initialized/');
        $this->board->markCell(0, 0);
    }

    public function testWhenAllCellGetsMarkedItIsMarked(): void
    {
        $anyValue = "42";
        $this->board = new BingoBoard(1, 1);
        $this->board->defineCell(0, 0, $anyValue);
        $this->board->markCell(0, 0);
        $this->assertTrue($this->board->is_marked(0, 0));
    }
}
